### ISSUE TITLE
Initialize the DB when the container starts

### ISSUE THREAD

**Issue by mafh314** [kubearchive/kubearchive#85]
Create the necessary tables and load the data for testing mounting .sql files under /docker-entrypoint-initdb.d 
(https://github.com/docker-library/docs/blob/master/postgres/README.md#initialization-scripts)

**Comment by rh-hemartin** [comment]
I think this will have to be manual in some way, I wouldn't push too much for this one. If we need to make a manual step, the manual step can be to apply those files and that it, no need to think the process to be able to mount the files before the container is created.

**Comment by rh-hemartin** [comment]
For the future, I think this should not live in the database side. I think the responsible should be the sink/api that runs the database initialization before starting the application, as any other database application does. This will ensure that the moment the sink/api is started, the database is in the correct state.

**Comment by maruiz93** [comment]
I wouldn't let the API (a reader with only read permissions on the DB) should be able to create schemas / tables.

I have my doubts on the sink too. I prefer to have everything decoupled, and I would say that both sink and api should check the connection to the db when starting and if fails fail to start the container.

I get your point though of not having the tables provided in the database when the pod starts because I think that the user of kubearchive may provide its own DB without the schema, just provide the permissions in a specific database.

I think that in a production environment the operator should have the responsability of setting the tables.

Now, if we accept that the operator would be the responsible of configuring the DB schema, as in development we aren't using it and we are using helm, the equivalent procedure would be one of the following (as I see it):

* Run the init_db script manually
* Mount the .sql files in the DB pod as suggested in this issue

**Comment by maruiz93** [comment]
> We are using helm to initialize the database?

I meant that we are using helm instead an operator. 

If we agree on the operator to be the one that configure the schema in the database  (I see that we haven't yet) we should do something in development for initializing it. The two options that I see are:

* Run the script manually (this is what we are doing now, and this is outside of the helm install command)
* Make the DB container to automatically set up the kubearchive schema (so this will be done as part of the helm install command)

> I'm not sure I agree with this. The operator, at least currently, isn't doing any sort of dependency configuration/installation.

It's not a problem that the operator doesn't do it right now but this is a good discussion to set up the expectations on what the operator should and shouldn't do in the future.

As I stated before, for me, configuring the schema makes sense to be done out from the sink/api, and as I see the operator as the "deployer" I also see it as "the one who configure every kubearchive component to get things working". I envision it as:

- [DB] Create the appropriate tables/index (and lookup data if we have it at any point)
- [Events] Create the APIServerSource CRs
- [Sink] Deploy the sink
- [REST API Server] Deploy the API Server

**Comment by maruiz93** [comment]
@ggallen what component should be creating the tables from your POV?

**Comment by rh-hemartin** [comment]
> I meant that we are using helm instead an operator.

This is not true as they are not interchangeable. The chart and the operator do two different things. The chart deploys our main components, and the operator deploys the ApiServerSource on the namespaces that require it. With #107 we will use the operator in development too. Also, we will use the helm chart in production too.

The table configuration (migrations) will be perform by the sink, as it is the responsible to write into it. We can make the API check if the schema is the correct one, or just start and error 500 out if the table is not there.

**Comment by skoved** [comment]
Couldn't we have initializing the database be an optional part of the helm chart? There could be some value like `database.init` that if set to `true` would cause the helm chart to add `XXX.sql` a file mounted by the database pod. For example, [the `test` namespace that the helm chart creates works like this](https://github.com/kubearchive/kubearchive/blob/main/charts/kubearchive/values.yaml#L16). if you remove this entry for `test` in `watchNamespaces` then the chart won't create and configure it for you

**Comment by skoved** [comment]
> I think the responsible should be the sink/api that runs the database initialization before starting the application, as any other database application does. This will ensure that the moment the sink/api is started, the database is in the correct state.

I would be worried about having the sink be responsible for initializing the database. With the way that it currently works, one replica is created, but there's not that would prevent you from increasing the number of replicas and with the way that cloudevents work, this would work perfectly fine. When an ApiServerSource sends a cloudevent to the sink, Kube DNS will replace the service url with the cluster IP of one of the replicas (which one depends on the Cluster's DNS configuration). If the sink becomes responsible for managing the database schema, this could introduce race conditions. I think that whatever method we use for installation/deployment of the database should handle initialization and maybe migrations as well. I could also see a world where cluster admins may need to perform manual steps if database migrations are required

**Comment by rh-hemartin** [comment]
Now that we are are talking about multi-cluster deployments with the inclusion of cluster_uid and cluster on the schema for the table, we have the very same problem with the operator too. There can be multiple operators being deployed on multiple clusters and we would get on the very same race condition. Maybe we can explore the option of leaving it as a manual step for the deployer to do it.

**Comment by maruiz93** [comment]
The discussion of who should be responsible of deploying the schema in the database is tabled for now but I think that we can still do one thing related to the initial goal of this issue, extract the SQL statements of `init_db.go` to a new file (maybe `ddl.sql`) so `init_db.go` just apply it the same way it does with `resource.sql`.

Maybe we can also rename `resource.sql` to `dml-example.sql` or something like that.

WDYT?

**Comment by maruiz93** [comment]
Revisiting this topic, I was thinking that we could still address this with some surrounding tweaks, please, let me know what do you think:

1. Move `ddl-resource.sql` to the chart conf directory (or create a link to it)
2. Declare in `charts/kubearchive/template/database/` a `ConfigMap` that loads the content of `ddl-resources.sql` if `.Values.global.production==false`
3. Mount in the database deployment this configmap to be in `/docker-entrypoint-initdb.d` so it's loaded on initialization if `.Values.global.production==false` or in `/ddl.sql` otherwise
4. If `.Values.global.production==true` set the credentials of the database as `root` `root` `password`
5. In the `hack/quick-install.sh` script, replace the call of `hack/database-install.sh` for the following:

```bash
helm template kubearchive charts/kubearchive -n databases \
    -s templates/database/database.yaml \
    --set "database.enabled=true" \
    --set "global.production=true" > /tmp/kubearchive-database.yaml
    
kubectl create namespace databases
kubectl apply --namespace databases -f /tmp/kubearchive-database.yaml

kubectl rollout status deployment --namespace=databases --timeout=30s
sleep 15  # Wait for PostgreSQL to be really ready

echo "CREATE USER kubearchive WITH ENCRYPTED PASSWORD 'P0stgr3sdbP@ssword';" | kubectl exec -i -n databases deploy/kubearchive-database -- psql -h localhost
echo "CREATE DATABASE kubearchive WITH OWNER kubearchive;" | kubectl exec -i -n databases deploy/kubearchive-database -- psql -h localhost
kubectl exec -i -n databases deploy/kubearchive-database -- psql -h localhost --user=kubearchive kubearchive -f /ddl.sql
```

After that I believe we can remove `hack/database-install.sh` because the functionality of that script will be already covered.

The advantages I see for the changes described are:
* Having the database resources declared just once (now the content of the YAMLs embedded in `hack/install-database.sh` and the content of `charts/kubearchive/templates/database/database.yaml` is almost the same)
* Reduce the amount of code for the quick installation process
* Keep a workflow similar to the one provided in the docs for the users to configure the database after the DB is deployed (i.e, creation of kubearchive user and database, and the application of the DDL)
* Remove the step of applying the `ddl-resource.sql` in the development process. (`DEVELOPMENT.md` file should be updated too)

P.S. As a side note we could possibly also improve the script by extracting the kubearchive username credentials and DB from `kubectl get secret -n kubearchive kubearchive-database-credentials -o jsonpath="{.data}"` to avoid duplication there too.

**Comment by ggallen** [comment]
> echo "CREATE USER kubearchive WITH ENCRYPTED PASSWORD 'P0stgr3sdbP@ssword';" | kubectl exec -i -n databases deploy/kubearchive-database -- psql -h localhost
> echo "CREATE DATABASE kubearchive WITH OWNER kubearchive;" | kubectl exec -i -n databases deploy/kubearchive-database -- psql -h localhost

These command should all be inside the database SQL file.  That file should entirely self-contained, so a user can just apply it using `psql` and get the entire database configured.

I haven't been following this issue closely, but it seems like things have become very complicated.

**Comment by ggallen** [comment]
As I look more closely, I think this issue should be closed.   Initializing the database should not be done by KubeArchive at all.

The database configuration and setup should all be done outside of KubeArchive. We are not database experts.  The only thing KubeArchive should require is the database connection information.

**Comment by maruiz93** [comment]
TL;DR Please, note that my suggestion is only to simplify what we already have in place regarding the database which is only for *testing* and *development*.

We currently have the [hack/install-database.sh](https://github.com/kubearchive/kubearchive/blob/main/hack/install-database.sh) for the [hack/quick-install.sh](https://github.com/kubearchive/kubearchive/blob/main/hack/quick-install.sh#L8) that is used in this [integration test](https://github.com/kubearchive/kubearchive/blob/main/.github/workflows/go_test_integration.yml#L50)

We also have a database in place for development purposes that is installed with the `helm install` command.

We are not configuring things for the users but we have to deploy and configure a database for development and testing purposes.

You can check https://github.com/kubearchive/kubearchive/pull/378 to see what are the instructions that we aim to give the users regarding their database configuration.

**Comment by rh-hemartin** [comment]
Hey, we have some different topics here, for 

> These commands should all be inside the database SQL file. That file should entirely self-contained, so a user can just apply it using psql and get the entire database configured.

I created #381 to discuss this.

For

> TL;DR Please, note that my suggestion is only to simplify what we already have in place regarding the database which is only for testing and development.

Given that our development and production deployments are different, we talked about CI being closer to production to catch more problems. So if we want to simplify, I would go into the other direction: making development more similar to the CI (a proxy for production), not the other way around.

**Comment by maruiz93** [comment]
> if we want to simplify, I would go into the other direction: making development more similar to the CI (a proxy for production), not the other way around.

What would be your suggestions?

I envision the output of the simplification by having:

* The database resources just in one place
* The default credentials for KubeArchive just in one place
* Prevent having embedded YAML definitions inside a script (for readability)

Either way, I think my suggestion keeps the CI similar to the production workflow:
* It deploys a database without a kubearchive user
* It needs to create the user, the database and apply the rest of the DDL afterwards

I just "reuse" the Helm stuff for deploying a standard db

**Comment by maruiz93** [comment]
If we are potentially including the creation of the kubearchive user and the database in the DDL, we could adjust the db helm template to be more similar to the one that is currently in `hack/install-database.sh` and deploy the DB with a root user and then apply the DDL mounting it

**Comment by rh-hemartin** [comment]
> What would be your suggestions?

I think the topic of approaching testing and development has enough content to move to another issue. Please open it and gather your proposal there. I will add my suggestions there.

**Comment by maruiz93** [comment]
> I think the topic of approaching testing and development in regards of the database setup has enough content to move it to another issue. Please open it and gather your proposal there. I will add my suggestions there.

Opened https://github.com/kubearchive/kubearchive/issues/390

**Comment by maruiz93** [comment]
I believe that given all the changes we have done related to the database, moving it outside of helm chart and treating it as an integration makes this issue as non-applicable anymore.

Feel free to reopen it otherwise.
